### ISSUE TITLE
XMLSerializer: Adding a namespaced child w/o prefix to a namespaced parent with prefix incorrect output

### ISSUE THREAD

**Issue by philschatz** [xmldom/xmldom#262]
Howdy! This snippet illustrates the problem.

We have an element `<a:foo>` that has a namespace prefix and a child `<bar>` that belongs to the same namespace but does not have a prefix.

We then create `<broken>` using `createElementNS` and append that to `<a:foo>` and then loses its namespace when it is serialized.

```js
const { DOMParser, XMLSerializer } = require('xmldom')

const str = `
<a:foo xmlns:a="AAA">
    <bar xmlns="AAA"/>
</a:foo>
`.trim()

const doc = new DOMParser().parseFromString(str)

const child = doc.createElementNS('AAA', 'broken')
doc.documentElement.appendChild(child)
console.log(new XMLSerializer().serializeToString(doc))
/* The output:

<a:foo xmlns:a="AAA">
    <bar xmlns="AAA"/>
    <broken/>
</a:foo>

*/
```

**Comment by karfau** [comment]
Thx for reporting.

Sounds quite critical, but I don't think the resulting XML string is "invalid".
Of course it doesn't match what you intended by appending the node.

I will need to investigate and see if it's easy to fix, if so it's possible that it can make it into the next release (still no timeline for it and there are some other labor intensive topics that need to go into it).

**Comment by karfau** [comment]
Just some notes from first analysis for later reference:
This is the output when running this code in the browser:
```
<a:foo xmlns:a="AAA">
    <bar xmlns="AAA"/>
    <a:broken/>
</a:foo>
```
since the related specs describe it this way:
- https://dom.spec.whatwg.org/#ref-for-dom-document-createelementns%E2%91%A0
- https://www.w3.org/TR/DOM-Level-2-Core/#core-ID-DocCrElNS

The `localName`, `nodeName` and `tagName` all are correctly `broken`, since no prefix was provided in the second parameter.
When just serializing that element the result is `<broken xmlns="AAA"/>`
But when serializing the whole document the namespaceUri `AAA` is already declared with the prefix `a` in the parent, so that one needs to be picked up and added as a prefix when serializing.

**Comment by karfau** [comment]
On further analysis I learned that I need to better understand the difference between the `bar` tag in the XML (which also has no prefix and the AAA namespace declared as the default) and declaring the `broken` element using the DOM API.
To me it currently looks as if the code that creates the `bar` element calls it with the same argument as you do in you sample, so potentially xmldom is currently not able treat those two tags differently when serializing them.
It would mean that any fix that would be applied to `broken` might also be applied to `bar`.

I also found out that xmldom currently uses a single list to keep track of the "visibleNamespaces" during serialization, which means that the default namespace that is declared by `bar` affects the namespaces that are "visible for `broken`" which I currently assume is another bug...

For me this indicates that we will currently not be able to fix it in the upcoming release.
